extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/DataServerQueryHandler.java

```diff
@@ -195,7 +200,36 @@ public <RowType, QueryType> DataServerQueryResult<RowType> fetchRowsFromDataServ
       }
     }
 
-    return new DataServerQueryResult<>(yielders, handedOffSegments, dataSource);
+    return new DataServerQueryResult<>(yielders, handedOffSegments, dataSourceName);
+  }
+
+  /**
+   * Transform {@link InputNumberDataSource}s, which are only understood by MSQ tasks, back into a
+   * {@link TableDataSource}.
+   */
+  private DataSource transformDatasource(DataSource dataSource)
+  {
+    if (dataSource instanceof InputNumberDataSource) {
+      InputNumberDataSource numberDataSource = (InputNumberDataSource) dataSource;
```

Comment:
This should also have a branch to check for `RestrictedInputNumberDataSource`. If it sees one, create a `RestrictedDataSource` wrapping a `TableDataSource` using the correct `Policy`. Please include a test for this. You don't need to include an end-to-end test; a unit test for this function specifically is OK. To make it more testable, consider making it `static`.